internal/commands/vaultsecrets/run/run.go

```diff
@@ -149,53 +150,56 @@ func getAllSecretsForEnv(opts *RunOpts) ([]string, error) {
 	}
 
 	result := os.Environ()
-	collisions := make(map[string]bool, 0)
+	collisions := make(map[string][]*models.Secrets20231128OpenSecret, 0)
 
 	for _, secret := range res.Payload.Secrets {
 		// we need to append results in case of duplicates we want secrets to override
 		switch {
 		case secret.RotatingVersion != nil:
 			for name, value := range secret.RotatingVersion.Values {
 				fmtName := fmt.Sprintf("%v_%v", strings.ToUpper(secret.Name), strings.ToUpper(name))
-				processCollisions(collisions, fmtName)
+				collisions[fmtName] = append(collisions[fmtName], secret)
 				result = append(result, fmt.Sprintf("%v=%v", fmtName, value))
 			}
 		case secret.DynamicInstance != nil:
 			for name, value := range secret.DynamicInstance.Values {
 				fmtName := fmt.Sprintf("%v_%v", strings.ToUpper(secret.Name), strings.ToUpper(name))
-				processCollisions(collisions, fmtName)
+				collisions[fmtName] = append(collisions[fmtName], secret)
 				result = append(result, fmt.Sprintf("%v=%v", fmtName, value))
 			}
 		case secret.StaticVersion != nil:
 			fmtName := strings.ToUpper(secret.Name)
-			processCollisions(collisions, fmtName)
+			collisions[fmtName] = append(collisions[fmtName], secret)
 			result = append(result, fmt.Sprintf("%v=%v", fmtName, secret.StaticVersion.Value))
 		}
 	}
 
-	// check collisions by dropping unique records from the map
-	for fmtName, collided := range collisions {
-		if !collided {
-			delete(collisions, fmtName)
-		} else {
-			_, err = fmt.Fprintf(opts.IO.Err(), "%s Environment variable \"%s\" was assigned more than once\n",
-				opts.IO.ColorScheme().WarningLabel(), fmtName)
+	// check collisions and emit an error for each
+	hasCollisions := false
+	for fmtName, uses := range collisions {
+		if len(uses) > 1 {
+			hasCollisions = true
+			var offenders string
+			for i, use := range uses {
+				offenders += fmt.Sprintf("\"%s\" [%s]", use.Name, use.Type)
+				if len(uses)-1 != i {
+					offenders += ", "
+				}
```

Comment:
nit: this can be a tiny bit cleaner if `offenders` is a `[]string` and we use `strings.Join(offenders, ", ")`

Reply:
🧠  3f10ceb